File: backend/app/schemas/operation.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import EntityRead
from app.schemas.work_order import OperationStatus, WorkOrderStatus


class DefectInput(BaseModel):
    reason_code: str = Field(min_length=1, max_length=64)
    qty: Decimal = Field(gt=0, max_digits=18, decimal_places=6)


class ActualMaterialInput(BaseModel):
    material_code: str = Field(min_length=1, max_length=64)
    qty: Decimal = Field(gt=0, max_digits=18, decimal_places=6)
    lot_no: str | None = Field(default=None, max_length=128)


class OperationStart(BaseModel):
    operator_code: str | None = Field(default=None, max_length=64)
    remark: str | None = None


class OperationStateChange(BaseModel):
    operator_code: str | None = Field(default=None, max_length=64)
    reason: str | None = None


class OperationClock(BaseModel):
    good_qty: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    bad_qty: Decimal = Field(ge=0, max_digits=18, decimal_places=6)
    defects: list[DefectInput] = Field(default_factory=list)
    actual_materials: list[ActualMaterialInput] = Field(default_factory=list)
    operator_code: str | None = Field(default=None, max_length=64)
    client_timestamp: datetime | None = None
    remark: str | None = None


class OperationRead(EntityRead):
    work_order_id: UUID
    work_order_no: str
    work_order_status: WorkOrderStatus
    material_code: str
    material_name: str
    seq: int
    operation_code: str
    operation_name: str
    work_center_id: UUID
    work_center_code: str
    work_center_name: str
    setup_time_sec: int
    unit_time_sec: int
    planned_duration_sec: int
    planned_qty: Decimal
    good_qty: Decimal
    bad_qty: Decimal
    status: OperationStatus
    started_at: datetime | None = None
    started_by_operator_code: str | None = None
    started_by_operator_name: str | None = None


class OperationClockRead(BaseModel):
    operation: OperationRead
    work_order_status: WorkOrderStatus
    next_operation_id: UUID | None = None
    clock_record_id: UUID

File: frontend/src/types/operation.ts
import type { EntityRead, UUID } from './masterData'
import type { OperationStatus, WorkOrderStatus } from './workOrder'

export interface OperationRead extends EntityRead {
  work_order_id: UUID
  work_order_no: string
  work_order_status: WorkOrderStatus
  material_code: string
  material_name: string
  seq: number
  operation_code: string
  operation_name: string
  work_center_id: UUID
  work_center_code: string
  work_center_name: string
  setup_time_sec: number
  unit_time_sec: number
  planned_duration_sec: number
  planned_qty: string
  good_qty: string
  bad_qty: string
  status: OperationStatus
  assigned_operator_code: string | null
  assigned_operator_name: string | null
  started_at: string | null
  started_by_operator_code: string | null
  started_by_operator_name: string | null
}

export interface OperationClockResponse {
  operation: OperationRead
  work_order_status: WorkOrderStatus
  next_operation_id: UUID | null
  clock_record_id: UUID
  elapsed_seconds: number | null
  time_anomaly: boolean
  time_anomaly_reason: string | null
  time_anomaly_detail: Record<string, unknown> | null
}

export type BackfillStatus = 'pending' | 'approved' | 'rejected'

export interface OperationBackfillPayload {
  started_at: string
  ended_at: string
  good_qty: string
  bad_qty: string
  defects: Array<{ reason_code: string; qty: string }>
  actual_materials: Array<{ material_code: string; qty: string; lot_no?: string | null }>
  operator_code?: string | null
  reason: string
  remark: string | null
}

export interface OperationBackfillReviewPayload {
  review_remark: string | null
}

export interface OperationBackfillRequestRead extends EntityRead {
  work_order_id: UUID
  operation_id: UUID
  clock_record_id: UUID | null
  work_order_no: string
  operation_seq: number
  operation_code: string
  operation_name: string
  work_center_id: UUID
  work_center_code: string
  work_center_name: string
  applicant_code: string
  applicant_name: string
  operator_code: string
  operator_name: string
  started_at: string
  ended_at: string
  elapsed_seconds: number
  good_qty: string
  bad_qty: string
  defects: Array<Record<string, unknown>>
  material_consumed: Array<Record<string, unknown>>
  reason: string
  remark: string | null
  status: BackfillStatus
  reviewed_by: string | null
  reviewed_at: string | null
  review_remark: string | null
}
